src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+
+        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private bool _solutionChanged;
+        private HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+
+        private readonly ConcurrentDictionary<ProjectId, Dictionary<string, string>> _projectToInstalledPackageAndVersion =
+            new ConcurrentDictionary<ProjectId, Dictionary<string, string>>();
+
+        public bool TryInstallPackage(Workspace workspace, ProjectId projectId, string packageName, string versionOpt)
+        {
+            this.AssertIsForeground();
+
+            if (workspace == _workspace && _workspace != null && _packageInstallerServices != null)
+            {
+                var dte = _workspace.GetVsService<SDTE, EnvDTE.DTE>();
+                var dteProject = _workspace.TryGetDTEProject(projectId);
+                if (dteProject != null)
+                {
+                    try
+                    {
+                        if (!_packageInstallerServices.IsPackageInstalled(dteProject, packageName))
+                        {
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0, packageName);
+                            _packageInstaller.InstallPackage(source: null, project: dteProject, packageId: packageName, version: versionOpt, ignoreDependencies: false);
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_completed, packageName);
+                        }
+
+                        return true;
+                    }
+                    catch
+                    {
+                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_failed, packageName);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        internal void Connect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            _workspace = workspace;
+
+            var componentModel = workspace.GetVsService<SComponentModel, IComponentModel>();
+            _packageInstallerServices = componentModel.GetService<IVsPackageInstallerServices>();
+            _packageInstaller = componentModel.GetService<IVsPackageInstaller>();
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        internal void Disconnect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            Debug.Assert(workspace == _workspace);
+            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+            _workspace = null;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            // Can be called on any thread.
+            bool localSolutionChanged = false;
+            ProjectId localChangedProject = null;
+            switch (e.Kind)
+            {
+                default:
```

Comment:
default at the top is uncommon, right? I almost left a comment saying you'd be calling to OnWorkspaceChangedOnForeground for unrelated changes. :)